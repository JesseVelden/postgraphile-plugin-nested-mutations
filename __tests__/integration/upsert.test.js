const { graphql } = require('graphql');
const { withSchema } = require('../helpers');

test(
  'ignore conflict when create is nested in upsert',
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );

      create table p.child (
        id serial primary key,
        parent_id integer,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );
      insert into p.parent values(1, 'test');
      insert into p.child values(99, 1, 'test child');
    `,
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          updateParentById(
            input: {
              id: 1
              parentPatch: {
                childrenUsingId: {
                  upsert: [{
                    create: { id: 99, name: "test child conflict" },
                  }]
                }
              }
            }
          ) {
            parent {
              id
              name
              childrenByParentId {
                nodes {
                  id
                  parentId
                  name
                }
              }
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');

      const data = result.data.updateParentById.parent;
      expect(data.childrenByParentId.nodes).toHaveLength(1);
      data.childrenByParentId.nodes.map((n) =>
        expect(n.parentId).toBe(data.id),
      );
    },
  }),
);

test(
  "deleteOthers removes everything that isn't connected or created in upsert",
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );

      create table p.child (
        id serial primary key,
        parent_id integer,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );
      insert into p.parent values(1, 'test');
      insert into p.child values(99, 1, 'child kept');
      insert into p.child values(100, 1, 'child that will be removed');
    `,
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          updateParentById(
            input: {
              id: 1
              parentPatch: {
                childrenUsingId: {
                  deleteOthers: true,
                  connectById: { id: 99 },
                  upsert: [{
                    create: { id: 101, name: "new child" },
                  }]
                }
              }
            }
          ) {
            parent {
              id
              name
              childrenByParentId {
                nodes {
                  id
                  parentId
                  name
                }
              }
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');

      const data = result.data.updateParentById.parent;
      expect(data.childrenByParentId.nodes).toHaveLength(2);
      data.childrenByParentId.nodes.map((n) =>
        expect(n.parentId).toBe(data.id),
      );
    },
  }),
);
