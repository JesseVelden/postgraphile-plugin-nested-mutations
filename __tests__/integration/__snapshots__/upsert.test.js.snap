// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteOthers removes everything that isn't connected or created in upsert 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "Node": Array [
      "Query",
      "Child",
      "Parent",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": Object {},
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyDelete": "ChildChildPkeyDelete",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdDelete": "ChildNodeIdDelete",
    "ChildOnChildForChildParentFkeyNodeIdUpdate": "ChildOnChildForChildParentFkeyNodeIdUpdate",
    "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate": "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate",
    "ChildParentFkeyChildCreateInput": "ChildParentFkeyChildCreateInput",
    "ChildParentFkeyChildUpsertInput": "ChildParentFkeyChildUpsertInput",
    "ChildParentFkeyInput": "ChildParentFkeyInput",
    "ChildParentFkeyInverseInput": "ChildParentFkeyInverseInput",
    "ChildParentFkeyParentCreateInput": "ChildParentFkeyParentCreateInput",
    "ChildParentFkeyParentUpsertInput": "ChildParentFkeyParentUpsertInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByIdInput": "DeleteChildByIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByIdInput": "DeleteParentByIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdDelete": "ParentNodeIdDelete",
    "ParentOnChildForChildParentFkeyNodeIdUpdate": "ParentOnChildForChildParentFkeyNodeIdUpdate",
    "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate": "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyDelete": "ParentParentPkeyDelete",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdateChildByIdInput": "UpdateChildByIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByIdInput": "UpdateParentByIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
    "updateChildOnChildForChildParentFkeyPatch": "updateChildOnChildForChildParentFkeyPatch",
    "updateParentOnChildForChildParentFkeyPatch": "updateParentOnChildForChildParentFkeyPatch",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;

exports[`ignore conflict when create is nested in upsert 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "Node": Array [
      "Query",
      "Child",
      "Parent",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": Object {},
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyDelete": "ChildChildPkeyDelete",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdDelete": "ChildNodeIdDelete",
    "ChildOnChildForChildParentFkeyNodeIdUpdate": "ChildOnChildForChildParentFkeyNodeIdUpdate",
    "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate": "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate",
    "ChildParentFkeyChildCreateInput": "ChildParentFkeyChildCreateInput",
    "ChildParentFkeyChildUpsertInput": "ChildParentFkeyChildUpsertInput",
    "ChildParentFkeyInput": "ChildParentFkeyInput",
    "ChildParentFkeyInverseInput": "ChildParentFkeyInverseInput",
    "ChildParentFkeyParentCreateInput": "ChildParentFkeyParentCreateInput",
    "ChildParentFkeyParentUpsertInput": "ChildParentFkeyParentUpsertInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByIdInput": "DeleteChildByIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByIdInput": "DeleteParentByIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdDelete": "ParentNodeIdDelete",
    "ParentOnChildForChildParentFkeyNodeIdUpdate": "ParentOnChildForChildParentFkeyNodeIdUpdate",
    "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate": "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyDelete": "ParentParentPkeyDelete",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdateChildByIdInput": "UpdateChildByIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByIdInput": "UpdateParentByIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
    "updateChildOnChildForChildParentFkeyPatch": "updateChildOnChildForChildParentFkeyPatch",
    "updateParentOnChildForChildParentFkeyPatch": "updateParentOnChildForChildParentFkeyPatch",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;
